File: scripts/generate_graph.py
""" This script generates all graphs that can be obtained
    from a reflexive Fano 3-tope by removing one vertex
    from it at a time.
    After collecting all the results, the complete graphs
    are saved in the file data/r3_graphs.txt
"""
import queue
from queue import Queue
import subprocess
from threading import Thread
from utils import hashdict, process_magma_output, \
    read_polytope_ids


# How many tasks to run concurrently.
n_workers = 4


def magma_worker(q, seen, Gs):
    """ Launches a magma process and parses its results. """
    while True:
        try:
            print("here!")
            n = q.get()
            print("got",n)
        except queue.Empty:
            return
        else:
            cmd = ["magma", "-b", f"rid:={n}", "generate_graph.m"]
            res = subprocess.run(cmd, capture_output=True)
            graph = process_magma_output(res.stdout)
# Update the ids seen so far.
# Important: the next three lines should not be replaced by:
#    seen |= graph.keys(), as this would not be atomic.
            newkeys = graph.keys() - seen
            for k in newkeys:
                seen.add(k) # All is fine: add is atomic.
# We need to make the dictionary hashable to store it in a set.
            imm_graph = hashdict(graph)
            Gs.add(imm_graph)
            q.task_done()


if __name__ == "__main__":
    pids = read_polytope_ids()
# Keep track of the polytopes already seen in some graph.
    seen = set()
# Set with the graphs for all the reflexive 3-topes.
    Gs = set()
    q = Queue(maxsize=n_workers)
    threads = [Thread(target=magma_worker, args=(q, seen, Gs), daemon=True) for _ in range(n_workers)]
    [t.start() for t in threads]
# We parse the database from the polytopes with more
# points to the polytopes with fewer points.
    for k in sorted(pids.keys(), reverse=True):
        for n in pids[k]:
            if n not in seen:
                q.put(n)

    q.join()

    with open("../data/r3_graphs.txt", "w") as fw:
        for G in Gs:
            for k, v in G.items():
                fw.write(f"{k}: {v}\n")
            fw.write("\n")

    print("All done.")

File: scripts/generate_graph.m
// Remove one element at a time from every reflexive Fano 3-tope P,
// and try to build a graph connecting all reflexive 3-topes
// that can be reached by removing one or more vertices from P.
import "utils.m": reflexive_id_for_polytope, is_equivalent_to_any,
    read_combinatorial_ids;


// We assume the called will execute this script with:
// magma -b rid:=<some int> ss:=true generate_graph.m
r_id := StringToInteger(rid);
single_step := StringToBoolean(ss);

// Read a pre-computed map from <#vertices, #edges, #facets, #points>
// to a reflexive id (useful for fast reverse searches).
comb_db := read_combinatorial_ids("../data/r3_cdata.txt");



// Vertices of the graph.
Vs := [Integers() | ];
// Edges of the graph. Maps a vertex id to a set of neighbours.
// The edges are directed from a polytope P to a polytope
// with one vertex less than P.
Es := AssociativeArray(Integers());
Es[r_id] := {Integers() | };

P := PolytopeReflexiveFanoDim3(r_id);
// Queue with the polytopes yet to consider.
// This queue stores pairs in the form:
// <polytope, id of closest reflexive ancestor>.
queue := [<P, r_id>];
seen := {Parent(P) | };  // Set of non-reflexive polytopes found so far.

while #queue gt 0 do
    P, m := Explode(queue[1]);
    vs := Vertices(P);
    for v in vs do
        vsnew := Exclude(vs, v);
        Q := Polytope(vsnew);
        if IsFano(Q) and IsReflexive(Q) then
            // We do not need to check the dimension.
            ok, n := reflexive_id_for_polytope(Q, comb_db);
            assert ok;
            if is_equivalent_to_any(Q, Vs) then
                Include(~Es[m], n);
            else
                if n in Vs then
                    Include(~Es[m], n);
                else
                    // Invariant: if n is in Vs, then n is also in Es.
                    Append(~queue, <Q, n>);
                    Append(~Vs, n);
                    Include(~Es[m], n);
                    Es[n] := {Integers() | };
                end if;
            end if;
        elif (not single_step)
            and (Dimension(Q) eq 3)
            and (not Q in seen) then
            // Q is not Fano or maybe not reflexive, but we should
            // keep removing vertices from Q and see if we can
            // find new reflexives.
            Append(~queue, <Q, m>);
            Include(~seen, Q);
        end if;
    end for;
    Remove(~queue, 1);
end while;

for id -> e in Es do
    printf "%o: %o\n", id, ToString(e);
end for;

quit;

